In [74]:
import pandas as pd
import plotly.express as px
import numpy as np

MINUTES_TO_HOURS = 1 / 60

# Bid Offer Analysis
This analysis focuses on March 19th 2022, based upon the analysis done in the other notebooks which identified this as a period of negative prices.

It focuses upon accepted offers to reduce generation.

In [6]:
df_bm_units = pd.read_excel("../data/BMUFuelType.xls", header=0)

In [7]:
df_fpn = pd.read_csv(
    "../data/PhysicalData_20220319_27.csv",
    skiprows=lambda x: x > 1541 or x == 0,
    names=[
        "Notification Type",
        "Unit ID",
        "Settlement Period",
        "From Time",
        "From Level",
        "To Time",
        "To Level",
    ],
    parse_dates=["From Time", "To Time"],
    date_parser=lambda x: pd.to_datetime(x, format="%Y%m%d%H%M%S"),
)  # This includes other kinds of data with different cols

df_fpn = (
    df_fpn.set_index("Unit ID")
    .join(df_bm_units.set_index("SETT_BMU_ID")["FUEL TYPE"])
    .rename(columns={"FUEL TYPE": "Fuel Type"})
)

There are many different forms here:
- FPN (Final Physical Notification): what the operators tell the grid they're going to generate
- QPN (Quiescent Physical Notifications)
- MEL (Maximum Export Limit)
- MIL (Maximum Import Limit)
- BOAL (Bid-Offer Acceptance Level Data)

In [8]:
df_boal = pd.read_csv(
    "../data/PhysicalData_20220319_27.csv",
    skiprows=lambda x: x < 5718,
    names=[
        "Notification Type",
        "Unit ID",
        "Accept ID",
        "Accept Time",
        "Deemed Flag",
        "SO Flag",
        "STOR Provider Flag",
        "RR Instruction Flag",
        "RR Schedule Flag",
        "From Time",
        "From Level",
        "To Time",
        "To Level",
    ],
    parse_dates=["Accept Time", "From Time", "To Time"],
    date_parser=lambda x: pd.to_datetime(x, format="%Y%m%d%H%M%S"),
)  # This includes other kinds of data with different cols

df_boal = df_boal[df_boal["Notification Type"] == "BOALF"]
df_boal["Accept ID"] = df_boal["Accept ID"].astype(int)
df_boal = (
    df_boal.set_index("Unit ID")
    .join(df_bm_units.set_index("SETT_BMU_ID")["FUEL TYPE"])
    .rename(columns={"FUEL TYPE": "Fuel Type"})
)

In [10]:
BOOL_COLS = [
    "Deemed Flag",
    "SO Flag",
    "STOR Provider Flag",
    "RR Instruction Flag",
    "RR Schedule Flag",
]

df_boal[BOOL_COLS] = df_boal[BOOL_COLS].apply(lambda x: x == "T")

I believe the SO Flag is identifying System Operator actions i.e. the ones we are interested in

In [11]:
df_boal[df_boal["SO Flag"]]

Notification Type  Accept ID         Accept Time  Deemed Flag  \
T_CARR-1              BOALF      79613 2022-03-19 12:09:00        False   
T_CARR-1              BOALF      79614 2022-03-19 12:44:00        False   
T_CARR-1              BOALF      79615 2022-03-19 13:26:00        False   
T_EAAO-1              BOALF       4379 2022-03-19 12:46:00        False   
T_EAAO-1              BOALF       4379 2022-03-19 12:46:00        False   
T_EAAO-1              BOALF       4380 2022-03-19 13:04:00        False   
T_LAGA-1              BOALF     133192 2022-03-19 12:00:00        False   
T_LAGA-1              BOALF     133193 2022-03-19 12:32:00        False   
T_MRWD-1              BOALF     144962 2022-03-19 12:01:00        False   
T_MRWD-1              BOALF     144965 2022-03-19 12:32:00        False   
T_PEMB-21             BOALF      88401 2022-03-19 12:01:00        False   
T_PEMB-21             BOALF      88402 2022-03-19 12:31:00        False   
T_WBURB-1             BOALF     112234 2022-03-19 12:08:00        False   
T_WBURB-1             BOALF     112234 2022-03-19 12:08:00        False   
T_WBURB-1             BOALF     112234 2022-03-19 12:08:00        False   
T_WBURB-1             BOALF     112234 2022-03-19 12:08:00        False   
T_WBURB-1             BOALF     112235 2022-03-19 12:44:00        False   
T_WBURB-1             BOALF     112236 2022-03-19 13:26:00        False   

           SO Flag  STOR Provider Flag  RR Instruction Flag  RR Schedule Flag  \
T_CARR-1      True               False                False             False   
T_CARR-1      True               False                False             False   
T_CARR-1      True               False                False             False   
T_EAAO-1      True               False                False             False   
T_EAAO-1      True               False                False             False   
T_EAAO-1      True               False                False             False   
T_LAGA-1      True               False                False             False   
T_LAGA-1      True               False                False             False   
T_MRWD-1      True               False                False             False   
T_MRWD-1      True               False                False             False   
T_PEMB-21     True               False                False             False   
T_PEMB-21     True               False                False             False   
T_WBURB-1     True               False                False             False   
T_WBURB-1     True               False                False             False   
T_WBURB-1     True               False                False             False   
T_WBURB-1     True               False                False             False   
T_WBURB-1     True               False                False             False   
T_WBURB-1     True               False                False             False   

                    From Time  From Level             To Time  To Level  \
T_CARR-1  2022-03-19 13:00:00       176.0 2022-03-19 13:30:00     176.0   
T_CARR-1  2022-03-19 12:46:00       176.0 2022-03-19 13:46:00     176.0   
T_CARR-1  2022-03-19 13:28:00       176.0 2022-03-19 14:13:00     176.0   
T_EAAO-1  2022-03-19 12:49:00       230.0 2022-03-19 13:19:00     230.0   
T_EAAO-1  2022-03-19 13:19:00       230.0 2022-03-19 13:22:00     336.0   
T_EAAO-1  2022-03-19 13:19:00       230.0 2022-03-19 13:49:00     230.0   
T_LAGA-1  2022-03-19 13:20:00         0.0 2022-03-19 13:30:00      17.0   
T_LAGA-1  2022-03-19 13:29:00        16.0 2022-03-19 14:00:00      71.0   
T_MRWD-1  2022-03-19 12:59:00       450.0 2022-03-19 13:30:00     450.0   
T_MRWD-1  2022-03-19 13:29:00       450.0 2022-03-19 14:00:00     450.0   
T_PEMB-21 2022-03-19 12:59:00       219.0 2022-03-19 13:30:00     219.0   
T_PEMB-21 2022-03-19 13:29:00       219.0 2022-03-19 14:00:00     219.0   
T_WBURB-1 2022-03-19 12:10:00       160.0 2022-03-19 13:10:00     160

In [12]:
df_boal_wind = df_boal[df_boal["Fuel Type"] == "WIND"]

In [13]:
df_boal_wind

Notification Type  Accept ID         Accept Time  Deemed Flag  \
E_BTUIW-3             BOALF       3642 2022-03-19 12:35:00        False   
E_BTUIW-3             BOALF       3643 2022-03-19 13:05:00        False   
T_AKGLW-3             BOALF       2005 2022-03-19 12:26:00        False   
T_AKGLW-3             BOALF       2006 2022-03-19 12:50:00        False   
T_AKGLW-3             BOALF       2007 2022-03-19 13:02:00        False   
...                     ...        ...                 ...          ...   
T_WLNYO-3             BOALF      22300 2022-03-19 13:07:00        False   
T_WLNYO-4             BOALF      23054 2022-03-19 12:30:00        False   
T_WLNYO-4             BOALF      23055 2022-03-19 12:50:00        False   
T_WLNYO-4             BOALF      23055 2022-03-19 12:50:00        False   
T_WLNYO-4             BOALF      23056 2022-03-19 13:10:00        False   

           SO Flag  STOR Provider Flag  RR Instruction Flag  RR Schedule Flag  \
E_BTUIW-3    False               False                False             False   
E_BTUIW-3    False               False                False             False   
T_AKGLW-3    False               False                False             False   
T_AKGLW-3    False               False                False             False   
T_AKGLW-3    False               False                False             False   
...            ...                 ...                  ...               ...   
T_WLNYO-3    False               False                False             False   
T_WLNYO-4    False               False                False             False   
T_WLNYO-4    False               False                False             False   
T_WLNYO-4    False               False                False             False   
T_WLNYO-4    False               False                False             False   

                    From Time  From Level             To Time  To Level  \
E_BTUIW-3 2022-03-19 13:20:00         0.0 2022-03-19 13:22:00      43.0   
E_BTUIW-3 2022-03-19 13:20:00         0.0 2022-03-19 13:50:00       0.0   
T_AKGLW-3 2022-03-19 12:58:00         0.0 2022-03-19 13:03:00      48.0   
T_AKGLW-3 2022-03-19 12:52:00         0.0 2022-03-19 13:52:00       0.0   
T_AKGLW-3 2022-03-19 13:04:00         0.0 2022-03-19 14:19:00       0.0   
...                       ...         ...                 ...       ...   
T_WLNYO-3 2022-03-19 13:17:00         0.0 2022-03-19 13:37:00       0.0   
T_WLNYO-4 2022-03-19 13:00:00         0.0 2022-03-19 13:06:00     253.0   
T_WLNYO-4 2022-03-19 13:00:00         0.0 2022-03-19 13:20:00       0.0   
T_WLNYO-4 2022-03-19 13:20:00         0.0 2022-03-19 13:26:00     262.0   
T_WLNYO-4 2022-03-19 13:20:00         0.0 2022-03-19 13:40:00       0.0   

          Fuel Type  
E_BTUIW-3      WIND  
E_BTUIW-3      WIND  
T_AKGLW-3      WIND  
T_AKGLW-3      WIND  
T_AKGLW-3      WIND  
...             ...  
T_WLNYO-3      WIND  
T_WLNYO-4      WIND  
T_WLNYO-4      WIND  
T_WLNYO-4      WIND  
T_WLNYO-4      WIND  

[100 rows x 13 columns]

# Identifying curtailment

Curtailment is if the BOALF level is less than the FPN

i.e. somebody is being paid to turn it down!

In [14]:
unit = "T_EAAO-1"
df_boal_wind.loc[unit]

Notification Type  Accept ID         Accept Time  Deemed Flag  \
T_EAAO-1             BOALF       4379 2022-03-19 12:46:00        False   
T_EAAO-1             BOALF       4379 2022-03-19 12:46:00        False   
T_EAAO-1             BOALF       4380 2022-03-19 13:04:00        False   
T_EAAO-1             BOALF       4381 2022-03-19 13:19:00        False   
T_EAAO-1             BOALF       4381 2022-03-19 13:19:00        False   

          SO Flag  STOR Provider Flag  RR Instruction Flag  RR Schedule Flag  \
T_EAAO-1     True               False                False             False   
T_EAAO-1     True               False                False             False   
T_EAAO-1     True               False                False             False   
T_EAAO-1    False               False                False             False   
T_EAAO-1    False               False                False             False   

                   From Time  From Level             To Time  To Level  \
T_EAAO-1 2022-03-19 12:49:00       230.0 2022-03-19 13:19:00     230.0   
T_EAAO-1 2022-03-19 13:19:00       230.0 2022-03-19 13:22:00     336.0   
T_EAAO-1 2022-03-19 13:19:00       230.0 2022-03-19 13:49:00     230.0   
T_EAAO-1 2022-03-19 13:21:00       230.0 2022-03-19 13:22:00     220.0   
T_EAAO-1 2022-03-19 13:22:00       220.0 2022-03-19 13:52:00     220.0   

         Fuel Type  
T_EAAO-1      WIND  
T_EAAO-1      WIND  
T_EAAO-1      WIND  
T_EAAO-1      WIND  
T_EAAO-1      WIND

In [15]:
df_fpn.loc[unit]

Notification Type                     PN
Settlement Period                     27
From Time            2022-03-19 13:00:00
From Level                         336.0
To Time              2022-03-19 13:30:00
To Level                           336.0
Fuel Type                           WIND
Name: T_EAAO-1, dtype: object

This is a curtailed unit! The PN was at 336 but it's being turned down to 220

In [16]:
def linearize_physical_data(df: pd.DataFrame):
    """Convert a From/To horizontal format to a long format with values at different timepoitns"""

    from_columns = ["From Level", "From Time"]
    to_columns = ["To Level", "To Time"]

    base_columns = [x for x in df.columns.copy() if x not in from_columns + to_columns]

    df = pd.concat(
        (
            df[base_columns + from_columns].rename(
                columns={"From Level": "Level", "From Time": "Time"}
            ),
            df[base_columns + to_columns].rename(
                columns={"To Level": "Level", "To Time": "Time"}
            ),
        )
    )

    df["Level"] = df["Level"].astype(float)
    return df

In [18]:
df_boal_linear = linearize_physical_data(df_boal.loc[unit])
df_boal_linear["Accept Time str"] = df_boal_linear["Accept Time"].astype(str)

px.line(df_boal_linear, "Time", "Level", color="Accept Time str")

Complexity here: the offers change over time, so I think you need to take the level indicated by the **latest** bid offer at every time point

We need to resample to a high frequency for each overlapping acceptance period separately and then recombine

In [51]:
def resolve_applied_bid_offer_level(df_linear: pd.DataFrame):
    """
    We can have multiple levels for a given timepoint, because levels are fixed
    at one point and then overwitten at a later timepoint, before the moment in
    question has arrived.

    We need to resolve them, choosing the latest possible commitment for each timepoint.

    We need to upsample the data first to achieve this.
    """

    out = []
    for accept_id, data in df_linear.groupby("Accept ID"):
        high_freq = (
            data.reset_index()
            .rename(columns={"index": "Unit"})
            .set_index("Time")
            .resample("T")
            .first()
        )
        out.append(high_freq.interpolate("ffill").fillna(method="ffill"))

    recombined = pd.concat(out)

    # Select the latest commitment for every timepoint
    resolved = recombined.reset_index().groupby("Time").last()

    return resolved

In [60]:
unit_boal_resolved = resolve_applied_bid_offer_level(df_boal_linear)
unit_boal_resolved.head()

Unit Notification Type  Accept ID  \
Time                                                         
2022-03-19 12:49:00  T_EAAO-1             BOALF     4379.0   
2022-03-19 12:50:00  T_EAAO-1             BOALF     4379.0   
2022-03-19 12:51:00  T_EAAO-1             BOALF     4379.0   
2022-03-19 12:52:00  T_EAAO-1             BOALF     4379.0   
2022-03-19 12:53:00  T_EAAO-1             BOALF     4379.0   

                            Accept Time  Deemed Flag  SO Flag  \
Time                                                            
2022-03-19 12:49:00 2022-03-19 12:46:00          0.0      1.0   
2022-03-19 12:50:00 2022-03-19 12:46:00          0.0      1.0   
2022-03-19 12:51:00 2022-03-19 12:46:00          0.0      1.0   
2022-03-19 12:52:00 2022-03-19 12:46:00          0.0      1.0   
2022-03-19 12:53:00 2022-03-19 12:46:00          0.0      1.0   

                     STOR Provider Flag  RR Instruction Flag  \
Time                                                           
2022-03-19 12:49:00                 0.0                  0.0   
2022-03-19 12:50:00                 0.0                  0.0   
2022-03-19 12:51:00                 0.0                  0.0   
2022-03-19 12:52:00                 0.0                  0.0   
2022-03-19 12:53:00                 0.0                  0.0   

                     RR Schedule Flag Fuel Type  Level      Accept Time str  
Time                                                                         
2022-03-19 12:49:00               0.0      WIND  230.0  2022-03-19 12:46:00  
2022-03-19 12:50:00               0.0      WIND  230.0  2022-03-19 12:46:00  
2022-03-19 12:51:00               0.0      WIND  230.0  2022-03-19 12:46:00  
2022-03-19 12:52:00               0.0      WIND  230.0  2022-03-19 12:46:00  
2022-03-19 12:53:00               0.0      WIND  230.0  2022-03-19 12:46:00

The FPN doesn't have multiple levels for a single time point so simply needs interpolating

In [61]:
unit_fpn_resolved = (
    linearize_physical_data(pd.DataFrame(df_fpn.loc[unit]).T)
    .set_index("Time")
    .resample("T")
    .mean()
    .interpolate()
)

In [62]:
unit_fpn_resolved["Notification Type"] = "FPN"
combined = pd.concat((resolved, unit_fpn_resolved))

The key graph: what did the unit actually do, relative to what it wanted to do?

In [64]:
px.line(combined, x=combined.index, y="Level", color="Notification Type")

In [77]:
def calculate_curtailment_in_mwh(unit_fpn_resolved, unit_boal_resolved) -> float:
    """
    Calculate the curtailment implied by the difference between FPN levels and BOAL

    """
    df_merged = unit_boal_resolved.join(unit_fpn_resolved, rsuffix="_FPN")
    df_merged["delta"] = df_merged["Level_FPN"] - df_merged["Level"]

    mw_minutes = df_merged["delta"].sum()

    return mw_minutes * MINUTES_TO_HOURS

In [78]:
calculate_curtailment_in_mwh(
    unit_fpn_resolved=unit_fpn_resolved, unit_boal_resolved=unit_boal_resolved
)

56.266666666666666

# TO DO
- Sanity check the numbers produced by this single analysis

- Wrap functionality so it can be run against a day's data, identifying curtailed units

- Why are some of these timepoints flagged as SO but others not?
